redis/connection.py

```diff
@@ -734,6 +715,188 @@ def _host_error(self):
+    ):
+        with self._cache_lock:
+            # Check if command response exists in a cache and it's not in progress.
+            if (
+                self._current_command_hash in self._cache
+                and self._cache[self._current_command_hash] != "caching-in-progress"
+            ):
+                return copy.deepcopy(self._cache[self._current_command_hash])
+
+        response = self._conn.read_response(
+            disable_decoding=disable_decoding,
+            disconnect_on_error=disconnect_on_error,
+            push_request=push_request,
+        )
+
+        with self._cache_lock:
+            # If response is None prevent from caching.
+            if response is None:
+                self._cache.pop(self._current_command_hash)
+                return response
+            # Prevent not-allowed command from caching.
+            elif self._current_command_hash is None:
+                return response
+
+            # Create separate mapping for keys
+            # or add current response to associated keys.
+            for key in self._current_command_keys:
+                if key in self._keys_mapping:
+                    if self._current_command_hash not in self._keys_mapping[key]:
+                        self._keys_mapping[key].append(self._current_command_hash)
+                else:
+                    self._keys_mapping[key] = [self._current_command_hash]
+
+            cache_entry = self._cache.get(self._current_command_hash, None)
+
+            # Cache only responses that still valid
+            # and wasn't invalidated by another connection in meantime.
+            if cache_entry is not None:
+                self._cache[self._current_command_hash] = response
+
+        return response
+
+    def pack_command(self, *args):
+        return self._conn.pack_command(*args)
+
+    def pack_commands(self, commands):
+        return self._conn.pack_commands(commands)
+
+    def _connect(self):
+        self._conn._connect()
+
+    def _host_error(self):
+        self._conn._host_error()
+
+    def _enable_tracking_callback(self, conn: ConnectionInterface) -> None:
+        conn.send_command("CLIENT", "TRACKING", "ON")
+        conn.read_response()
+        conn._parser.set_invalidation_push_handler(self._on_invalidation_callback)
+
+    def _process_pending_invalidations(self):
+        while self.retry.call_with_retry_on_false(lambda: self.can_read()):
+            self._conn.read_response(push_request=True)
+
+    def _on_invalidation_callback(self, data: List[Union[str, Optional[List[str]]]]):
+        with self._cache_lock:
+            # Flush cache when DB flushed on server-side
+            if data[1] is None:
+                self._cache.clear()
+            else:
+                for key in data[1]:
+                    normalized_key = ensure_string(key)
+                    if normalized_key in self._keys_mapping:
```

Comment:
It looks strange that we share cache across connections but don't do the same for `_keys_mapping`. It will lead to partial cache cleanup on invalidation when the same key was accessed by multiple connections, isn't it?